pkg/storage/ingest/reader.go

```diff
@@ -842,41 +864,43 @@ func (r *concurrentFetchers) pollFetches(ctx context.Context) (result kgo.Fetche
 	}
 }
 
-func (r *concurrentFetchers) fetchSingle(ctx context.Context, w fetchWant, _ log.Logger) (_ kgo.FetchPartition, fetchedBytes int) {
+func (r *concurrentFetchers) fetchSingle(ctx context.Context, w fetchWant) fetchResult {
 	req := kmsg.NewFetchRequest()
+	req.MinBytes = 1
+	req.Version = 13
+	req.MaxWaitMillis = 10000
+	req.MaxBytes = w.maxBytes()
 	req.Topics = []kmsg.FetchRequestTopic{{
 		Topic:   r.topicName,
 		TopicID: r.topicID,
 		Partitions: []kmsg.FetchRequestTopicPartition{{
-			Partition:          r.partitionID,
-			FetchOffset:        w.startOffset,
-			LastFetchedEpoch:   -1,
+			Partition:   r.partitionID,
+			FetchOffset: w.startOffset,
+			// CurrentLeaderEpoch: -1 means we don't know. Kafka brokers are ok with that.
+			// It does mean that we might end up fetching from an out-of-sync replica.
+			// If we provide this the broker would check if we have up-to-date data.
 			CurrentLeaderEpoch: -1,
```

Comment:
this is probably the shoddiest of this implementation. Do we want to deal with it?

Reply:
update: I found some implementation notes in franz-go 

https://github.com/grafana/mimir/blob/04df05d32320cd19281591a57c19cceba63ceabf/vendor/github.com/twmb/franz-go/pkg/kgo/topics_and_partitions.go#L530-L537

and the list of all possible errors. I think we should handle all of them as best as we can. Currently we handle only the most common one.

https://github.com/grafana/mimir/blob/04df05d32320cd19281591a57c19cceba63ceabf/vendor/github.com/twmb/franz-go/pkg/kmsg/generated.go#L4705-L4745

looks like a lot of code we'll have to write and test :(